File: Tools/scrape_epochhead_items.py
#!/usr/bin/env python3
"""Scrape EpochHead all-item listing and item tooltip green lines."""

from __future__ import annotations

from concurrent.futures import ThreadPoolExecutor, as_completed
from html import unescape
from pathlib import Path
import argparse
import json
import re
import subprocess
import time


ROOT = Path(__file__).resolve().parents[1]
CACHE = ROOT / "Tools" / "cache" / "epochhead_items"
OUT = CACHE / "items.json"
LIST_CACHE = CACHE / "list.json"
SNAPSHOT = ROOT / "SourceData" / "EpochHead" / "items"
SNAPSHOT_OUT = SNAPSHOT / "items.json"
SNAPSHOT_LIST = SNAPSHOT / "list.json"
BASE = "https://epochhead.com"
LIST_URL = BASE + "/items"
USER_AGENT = "EpochCN-all-item-sync/0.1"

ITEM_ROW_RE = re.compile(
    r'data-item-id="(?P<id>\d+)".{0,700}?'
    r'alt="(?P<name>[^"]+)".{0,1400}?'
    r"<td>(?P<category>.*?)</td><td>(?P<slot>.*?)</td><td[^>]*>(?P<ilevel>.*?)</td>",
    re.S,
)
PAGE_RE = re.compile(r"Page\s+<!-- -->\d+<!-- -->\s+of\s+<!-- -->(\d+)")
GREEN_RE = re.compile(r'<li class="text-green-400">(.*?)</li>', re.S)
TOOLTIP_UL_RE = re.compile(r'<ul class="list-none space-y-1">(.*?)</ul>', re.S)
TOOLTIP_LI_RE = re.compile(r'<li class="([^"]*)">(.*?)</li>', re.S)
TITLE_RE = re.compile(r"<title[^>]*>(.*?)</title>", re.S | re.I)
META_DESC_RE = re.compile(r'<meta\s+name="description"\s+content="([^"]*)"', re.S | re.I)


def fetch(url: str, retries: int = 3) -> str:
    last_error: Exception | None = None
    for attempt in range(retries):
        try:
            proc = subprocess.run(
                ["curl", "-L", "-sS", "--max-time", "30", "-A", USER_AGENT, url],
                check=True,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
            )
            return proc.stdout.decode("utf-8", "replace")
        except Exception as exc:
            last_error = exc
            time.sleep(0.75 + attempt)
    raise RuntimeError(f"fetch failed for {url}: {last_error}")


def clean_html(value: str) -> str:
    value = re.sub(r"<script[\s\S]*?</script>", " ", value or "", flags=re.I)
    value = re.sub(r"<style[\s\S]*?</style>", " ", value, flags=re.I)
    value = re.sub(r"<!--[\s\S]*?-->", " ", value)
    value = re.sub(r"<.*?>", " ", value)
    value = value.replace("\\u0026", "&")
    value = unescape(value)
    return re.sub(r"\s+", " ", value).strip()


def parse_list_page(html: str) -> list[dict]:
    rows: list[dict] = []
    for match in ITEM_ROW_RE.finditer(html):
        item_id = int(match.group("id"))
        rows.append({
            "id": item_id,
            "name": clean_html(match.group("name")),
            "category": clean_html(match.group("category")),
            "slot": clean_html(match.group("slot")),
            "ilevel": clean_html(match.group("ilevel")),
            "url": f"{BASE}/?item={item_id}",
        })
    return rows


def parse_page_count(html: str) -> int:
    match = PAGE_RE.search(html)
    return int(match.group(1)) if match else 1


def parse_title(html: str, fallback: str) -> str:
    match = TITLE_RE.search(html)
    if not match:
        return fallback
    title = clean_html(match.group(1))
    return re.sub(r"\s+—\s+EpochHead.*$", "", title).strip() or fallback


def parse_meta_description(html: str) -> str:
    match = META_DESC_RE.search(html)
    return clean_html(match.group(1)) if match else ""


def parse_green_lines(html: str) -> list[str]:
    lines: list[str] = []
    seen: set[str] = set()
    for match in GREEN_RE.finditer(html):
        line = clean_html(match.group(1))
        if line and line not in seen:
            seen.add(line)
            lines.append(line)
    return lines


def tooltip_color(class_name: str) -> str:
    if "text-green-400" in class_name:
        return "green"
    if "text-yellow" in class_name or "text-amber" in class_name:
        return "yellow"
    if "text-gray" in class_name or "text-neutral" in class_name or "muted" in class_name:
        return "gray"
    if "text-red" in class_name:
        return "red"
    return "normal"


def classify_tooltip_text(text: str) -> str:
    if re.match(r"^(Use|Equip|Chance on hit):", text):
        return "green"
    if re.match(r'^"[^"]+"$', text):
        return "yellow"
    return "normal"


def parse_tooltip_lines(html: str) -> list[dict]:
    match = TOOLTIP_UL_RE.search(html)
    if not match:
        return []
    rows: list[dict] = []
    for line_match in TOOLTIP_LI_RE.finditer(match.group(1)):
        class_name = line_match.group(1) or ""
        text = clean_html(line_match.group(2))
        if text:
            rows.append({"text": text, "color": tooltip_color(class_name), "className": class_name})
    if any(re.search(r"Retrieving item information", row["text"], re.I) for row in rows):
        return []
    return rows


def parse_meta_tooltip(title: str, description: str) -> list[dict]:
    text = clean_html(description or "")
    if title and text.startswith(title):
        text = clean_html(text[len(title):])
    if not text or re.search(r"Retrieving item information", text, re.I):
        return []

    marker = r'(?=\s+(?:"|Use:|Equip:|Chance on hit:|Binds|Requires|Unique|Races:|Classes:|Item Level|Speed|[0-9,.]+ - [0-9,.]+ Damage|\([0-9,.]+ damage per second\)|\+[0-9,.]+ )\b|$)'
    pattern = re.compile("|".join([
        r'"[^"]*"',
        r"Use:.*?" + marker,
        r"Equip:.*?" + marker,
        r"Chance on hit:.*?" + marker,
        r"Binds (?:to account|when picked up|when equipped|when used)",
        r"Unique(?:-Equipped)?(?: \(\d+\))?",
        r"One-Hand|Two-Hand|Main Hand|Off Hand|Held In Off-hand|Ranged|Thrown",
        r"Speed [0-9.]+",
        r"[0-9,.]+ - [0-9,.]+ Damage",
        r"\([0-9,.]+ damage per second\)",
        r"\+[0-9,.]+ [A-Za-z ]+",
        r"Requires Level \d+",
        r"Requires .*?" + marker,
        r"Item Level \d+",
        r"Races:.*?" + marker,
        r"Classes:.*?" + marker,
    ]), re.I)

    rows: list[dict] = []
    seen: set[str] = set()
    for match in pattern.finditer(text):
        line = clean_html(match.group(0))
        if line and line not in seen:
            seen.add(line)
            rows.append({"text": line, "color": classify_tooltip_text(line), "className": ""})
    return rows


def fetch_detail(item: dict) -> dict:
    html = fetch(item["url"])
    item = dict(item)
    item["name"] = parse_title(html, item.get("name") or "")
    item["description"] = parse_meta_description(html)
    tooltip = parse_tooltip_lines(html)
    if not tooltip:
        tooltip = parse_meta_tooltip(item["name"], item["description"])
    item["tooltip"] = tooltip
    green = [row["text"] for row in tooltip if row.get("color") == "green"]
    for line in parse_green_lines(html):
        if line not in green:
            green.append(line)
    item["green"] = green
    return item


def collect_list(total_pages_override: int | None = None) -> list[dict]:
    first = fetch(LIST_URL)
    total_pages = total_pages_override or parse_page_count(first)
    found: dict[int, dict] = {}
    for item in parse_list_page(first):
        found[item["id"]] = item
    print(f"list page 1/{total_pages}: items={len(found)}", flush=True)

    def fetch_list_page(page: int) -> tuple[int, str]:
        return page, fetch(f"{LIST_URL}?page={page}")

    with ThreadPoolExecutor(max_workers=10) as executor:
        futures = [executor.submit(fetch_list_page, page) for page in range(2, total_pages + 1)]
        for index, future in enumerate(as_completed(futures), 2):
            _, html = future.result()
            for item in parse_list_page(html):
                found[item["id"]] = item
            if index % 25 == 0 or index == total_pages:
                print(f"list pages {index}/{total_pages}: items={len(found)}", flush=True)

    items = [found[item_id] for item_id in sorted(found)]
    LIST_CACHE.write_text(json.dumps(items, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    return items


def load_existing_details() -> dict[int, dict]:
    details: dict[int, dict] = {}
    for path in (SNAPSHOT_OUT, OUT):
        if not path.exists():
            continue
        try:
            rows = json.loads(path.read_text(encoding="utf-8"))
        except json.JSONDecodeError:
            continue
        details.update({int(row["id"]): row for row in rows if row.get("id") and not row.get("error")})
    return details


def write_details(rows: list[dict]) -> None:
    text = json.dumps(rows, ensure_ascii=False, indent=2) + "\n"
    OUT.write_text(text, encoding="utf-8")
    SNAPSHOT.mkdir(parents=True, exist_ok=True)
    SNAPSHOT_OUT.write_text(text, encoding="utf-8")


def write_list(rows: list[dict]) -> None:
    text = json.dumps(rows, ensure_ascii=False, indent=2) + "\n"
    LIST_CACHE.write_text(text, encoding="utf-8")
    SNAPSHOT.mkdir(parents=True, exist_ok=True)
    SNAPSHOT_LIST.write_text(text, encoding="utf-8")


def main() -> None:
    parser = argparse.ArgumentParser(description="Scrape all EpochHead item details.")
    parser.add_argument("--pages", type=int, default=None, help="Override list page count for testing.")
    parser.add_argument("--reuse-list", action="store_true", help="Reuse cached item list if present.")
    parser.add_argument("--workers", type=int, default=16)
    args = parser.parse_args()

    CACHE.mkdir(parents=True, exist_ok=True)
    if args.reuse_list and LIST_CACHE.exists():
        items = json.loads(LIST_CACHE.read_text(encoding="utf-8"))
    elif args.reuse_list and SNAPSHOT_LIST.exists():
        items = json.loads(SNAPSHOT_LIST.read_text(encoding="utf-8"))
    else:
        items = collect_list(args.pages)
    if items:
        write_list(items)

    details = load_existing_details()
    todo = [item for item in items if int(item["id"]) not in details or not details[int(item["id"])].get("tooltip")]
    print(f"detail queue: {len(todo)}/{len(items)}", flush=True)

    with ThreadPoolExecutor(max_workers=max(1, args.workers)) as executor:
        futures = {executor.submit(fetch_detail, item): item for item in todo}
        for index, future in enumerate(as_completed(futures), 1):
            item = futures[future]
            try:
                row = future.result()
            except Exception as exc:
                row = dict(item)
                row["green"] = []
                row["error"] = str(exc)
            details[int(row["id"])] = row
            if index % 100 == 0 or index == len(todo):
                print(f"detail {index}/{len(todo)} complete={len(details)}/{len(items)}", flush=True)
                rows = [details[item_id] for item_id in sorted(details)]
                write_details(rows)

    rows = [details[item_id] for item_id in sorted(details)]
    write_details(rows)
    green_items = sum(1 for row in rows if row.get("green"))
    print(f"wrote {OUT}", flush=True)
    print(f"items={len(rows)} green_items={green_items}", flush=True)


if __name__ == "__main__":
    main()

File: Tools/scrape_epochhead_items.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const path = require("path");

const ROOT = path.resolve(__dirname, "..");
const CACHE = path.join(ROOT, "Tools", "cache", "epochhead_items");
const OUT = path.join(CACHE, "items.json");
const LIST_CACHE = path.join(CACHE, "list.json");
const SNAPSHOT = path.join(ROOT, "SourceData", "EpochHead", "items");
const SNAPSHOT_OUT = path.join(SNAPSHOT, "items.json");
const SNAPSHOT_LIST = path.join(SNAPSHOT, "list.json");
const SNAPSHOT_MANIFEST = path.join(SNAPSHOT, "manifest.json");
const BASE = "https://epochhead.com";
const LIST_URL = `${BASE}/items`;
const USER_AGENT = "EpochCN-all-item-sync/0.2";
let requestTimeoutMs = 12000;
let requestRetries = 2;
let requestDelayMs = 0;

const ITEM_ROW_RE = /data-item-id="(?<id>\d+)"[\s\S]{0,700}?alt="(?<name>[^"]+)"[\s\S]{0,1400}?<td>(?<category>[\s\S]*?)<\/td><td>(?<slot>[\s\S]*?)<\/td><td[^>]*>(?<ilevel>[\s\S]*?)<\/td>/g;
const PAGE_RE = /Page\s+<!-- -->\d+<!-- -->\s+of\s+<!-- -->(\d+)/;
const GREEN_RE = /<li class="text-green-400">([\s\S]*?)<\/li>/g;
const TOOLTIP_UL_RE = /<ul class="list-none space-y-1">([\s\S]*?)<\/ul>/;
const TOOLTIP_LI_RE = /<li class="([^"]*)">([\s\S]*?)<\/li>/g;
const TITLE_RE = /<title[^>]*>([\s\S]*?)<\/title>/i;
const META_DESC_RE = /<meta\s+name="description"\s+content="([^"]*)"/i;

function argValue(name, fallback) {
  const index = process.argv.indexOf(name);
  if (index >= 0 && process.argv[index + 1]) return process.argv[index + 1];
  return fallback;
}

function hasArg(name) {
  return process.argv.includes(name);
}

function decodeEntities(text) {
  return String(text || "")
    .replace(/\\u0026/g, "&")
    .replace(/&quot;/g, '"')
    .replace(/&#39;|&apos;/g, "'")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&amp;/g, "&")
    .replace(/&#(\d+);/g, (_, n) => String.fromCodePoint(Number(n)))
    .replace(/&#x([0-9a-f]+);/gi, (_, n) => String.fromCodePoint(parseInt(n, 16)));
}

function cleanHtml(value) {
  return decodeEntities(value)
    .replace(/<script[\s\S]*?<\/script>/gi, " ")
    .replace(/<style[\s\S]*?<\/style>/gi, " ")
    .replace(/<!--[\s\S]*?-->/g, " ")
    .replace(/<.*?>/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

async function fetchText(url, retries = requestRetries) {
  let lastError = null;
  for (let attempt = 0; attempt < retries; attempt += 1) {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), requestTimeoutMs);
    try {
      const response = await fetch(url, {
        headers: { "user-agent": USER_AGENT },
        redirect: "follow",
        signal: controller.signal,
      });
      if (!response.ok) throw new Error(`HTTP ${response.status}`);
      return await response.text();
    } catch (error) {
      lastError = error;
      await new Promise(resolve => setTimeout(resolve, 400 + attempt * 800));
    } finally {
      clearTimeout(timer);
    }
  }
  throw new Error(`fetch failed for ${url}: ${lastError && lastError.message}`);
}

async function politeDelay(workerIndex) {
  if (requestDelayMs <= 0) return;
  const jitter = (workerIndex % 7) * 137;
  await new Promise(resolve => setTimeout(resolve, requestDelayMs + jitter));
}

function parseListPage(html) {
  const rows = [];
  for (const match of html.matchAll(ITEM_ROW_RE)) {
    const itemId = Number(match.groups.id);
    rows.push({
      id: itemId,
      name: cleanHtml(match.groups.name),
      category: cleanHtml(match.groups.category),
      slot: cleanHtml(match.groups.slot),
      ilevel: cleanHtml(match.groups.ilevel),
      url: `${BASE}/?item=${itemId}`,
    });
  }
  return rows;
}

function parsePageCount(html) {
  const match = PAGE_RE.exec(html);
  return match ? Number(match[1]) : 1;
}

function parseTitle(html, fallback) {
  const match = TITLE_RE.exec(html);
  if (!match) return fallback || "";
  return cleanHtml(match[1]).replace(/\s+—\s+EpochHead.*$/, "").trim() || fallback || "";
}

function parseMetaDescription(html) {
  const match = META_DESC_RE.exec(html);
  return match ? cleanHtml(match[1]) : "";
}

function parseGreenLines(html) {
  const seen = new Set();
  const lines = [];
  for (const match of html.matchAll(GREEN_RE)) {
    const line = cleanHtml(match[1]);
    if (line && !seen.has(line)) {
      seen.add(line);
      lines.push(line);
    }
  }
  return lines;
}

async function collectList(pageOverride) {
  const first = await fetchText(LIST_URL);
  const totalPages = pageOverride || parsePageCount(first);
  const found = new Map();
  for (const item of parseListPage(first)) found.set(item.id, item);
  console.log(`list page 1/${totalPages}: items=${found.size}`);

  const pages = [];
  for (let page = 2; page <= totalPages; page += 1) pages.push(page);
  let done = 1;
  await runQueue(pages, 12, async page => {
    const html = await fetchText(`${LIST_URL}?page=${page}`);
    for (const item of parseListPage(html)) found.set(item.id, item);
    done += 1;
    if (done % 25 === 0 || done === totalPages) {
      console.log(`list pages ${done}/${totalPages}: items=${found.size}`);
    }
  });

  const rows = [...found.entries()].sort((a, b) => a[0] - b[0]).map(([, item]) => item);
  writeJson(LIST_CACHE, rows);
  return rows;
}

function loadJson(file, fallback) {
  if (!fs.existsSync(file)) return fallback;
  try {
    return JSON.parse(fs.readFileSync(file, "utf8"));
  } catch {
    return fallback;
  }
}

function writeJson(file, rows) {
  fs.mkdirSync(path.dirname(file), { recursive: true });
  const tmp = `${file}.tmp`;
  fs.writeFileSync(tmp, `${JSON.stringify(rows, null, 2)}\n`);
  fs.renameSync(tmp, file);
}

function sha256(file) {
  const crypto = require("crypto");
  return crypto.createHash("sha256").update(fs.readFileSync(file)).digest("hex");
}

function writeSnapshotManifest() {
  if (!fs.existsSync(SNAPSHOT_LIST) || !fs.existsSync(SNAPSHOT_OUT)) return;
  const list = loadJson(SNAPSHOT_LIST, []);
  const items = loadJson(SNAPSHOT_OUT, []);
  const manifest = {
    source: LIST_URL,
    detailUrlPattern: `${BASE}/?item={id}`,
    capturedAt: new Date().toISOString().slice(0, 10),
    list: {
      file: "list.json",
      count: list.length,
      sha256: sha256(SNAPSHOT_LIST),
    },
    details: {
      file: "items.json",
      count: items.length,
      greenItemCount: items.filter(row => row.green && row.green.length).length,
      errorCount: items.filter(row => row.error).length,
      sha256: sha256(SNAPSHOT_OUT),
    },
    schema: {
      list: ["id", "name", "category", "slot", "ilevel", "url"],
      details: ["id", "name", "category", "slot", "ilevel", "url", "description", "tooltip", "green"],
    },
    notes: [
      "list.json is the complete 335-page EpochHead item index captured from /items.",
      "items.json is an incremental detail snapshot. It is intentionally committed so future localization work can reuse it without re-scraping completed items.",
      "Rows with fetch errors are excluded; missing detail rows should be resumed by Tools/scrape_epochhead_items.js.",
    ],
  };
  writeJson(SNAPSHOT_MANIFEST, manifest);
}

function writeDurableJson(cacheFile, snapshotFile, rows) {
  writeJson(cacheFile, rows);
  writeJson(snapshotFile, rows);
  writeSnapshotManifest();
}

function loadExistingDetails() {
  const merged = new Map();
  for (const rows of [loadJson(SNAPSHOT_OUT, []), loadJson(OUT, [])]) {
    for (const row of rows) {
      if (row && row.id && !row.error) merged.set(Number(row.id), row);
    }
  }
  const rows = [...merged.entries()].sort((a, b) => a[0] - b[0]).map(([, row]) => row);
  const details = new Map();
  for (const row of rows) {
    if (row && row.id && !row.error) details.set(Number(row.id), row);
  }
  return details;
}

function classifyTooltipText(text) {
  if (/^\(\d+\)\s*Set:/.test(text)) return "green";
  if (/^(Use|Equip|Chance on hit):/.test(text)) return "green";
  if (/^"[^"]+"$/.test(text)) return "yellow";
  return "normal";
}

function tooltipColor(className) {
  if (/text-green-400/.test(className)) return "green";
  if (/text-yellow|text-amber/.test(className)) return "yellow";
  if (/text-gray|text-neutral|muted/.test(className)) return "gray";
  if (/text-red/.test(className)) return "red";
  return "normal";
}

function parseTooltipLines(html) {
  const match = TOOLTIP_UL_RE.exec(html);
  if (!match) return [];
  const result = [];
  for (const lineMatch of match[1].matchAll(TOOLTIP_LI_RE)) {
    const className = lineMatch[1] || "";
    const text = cleanHtml(lineMatch[2]);
    if (!text) continue;
    result.push({
      text,
      color: tooltipColor(className),
      className,
    });
  }
  if (result.some(line => /Retrieving item information/i.test(line.text))) return [];
  return result;
}

function parseMetaTooltip(title, description) {
  let text = cleanHtml(description || "");
  if (title && text.startsWith(title)) text = cleanHtml(text.slice(title.length));
  if (!text || /Retrieving item information/i.test(text)) return [];

  const marker = String.raw`(?=\s+(?:"|Use:|Equip:|Chance on hit:|Binds|Requires|Unique|Races:|Classes:|Item Level|Speed|[0-9,.]+ - [0-9,.]+ Damage|\([0-9,.]+ damage per second\)|\+[0-9,.]+ )\b|$)`;
  const pattern = new RegExp([
    String.raw`"[^"]*"`,
    String.raw`Use:.*?` + marker,
    String.raw`Equip:.*?` + marker,
    String.raw`Chance on hit:.*?` + marker,
    String.raw`Binds (?:to account|when picked up|when equipped|when used)`,
    String.raw`Unique(?:-Equipped)?(?: \(\d+\))?`,
    String.raw`One-Hand|Two-Hand|Main Hand|Off Hand|Held In Off-hand|Ranged|Thrown`,
    String.raw`Speed [0-9.]+`,
    String.raw`[0-9,.]+ - [0-9,.]+ Damage`,
    String.raw`\([0-9,.]+ damage per second\)`,
    String.raw`\+[0-9,.]+ [A-Za-z ]+`,
    String.raw`Requires Level \d+`,
    String.raw`Requires .*?` + marker,
    String.raw`Item Level \d+`,
    String.raw`Races:.*?` + marker,
    String.raw`Classes:.*?` + marker,
  ].join("|"), "gi");

  const rows = [];
  const seen = new Set();
  let match;
  while ((match = pattern.exec(text))) {
    const line = cleanHtml(match[0]);
    if (line && !seen.has(line)) {
      seen.add(line);
      rows.push({ text: line, color: classifyTooltipText(line), className: "" });
    }
  }
  return rows;
}

async function fetchDetail(item) {
  const html = await fetchText(item.url);
  const title = parseTitle(html, item.name || "");
  const description = parseMetaDescription(html);
  let tooltip = parseTooltipLines(html);
  if (!tooltip.length) tooltip = parseMetaTooltip(title, description);
  return {
    ...item,
    name: title,
    description,
    tooltip,
    green: tooltip.filter(line => line.color === "green" || classifyTooltipText(line.text) === "green").map(line => line.text).concat(parseGreenLines(html))
      .filter((line, index, lines) => line && lines.indexOf(line) === index),
  };
}

async function runQueue(items, workers, task) {
  let next = 0;
  async function worker(workerIndex) {
    while (next < items.length) {
      const index = next;
      next += 1;
      await task(items[index], index, workerIndex);
    }
  }
  await Promise.all(Array.from({ length: Math.max(1, workers) }, (_, index) => worker(index)));
}

async function main() {
  const workers = Number(argValue("--workers", "64"));
  requestTimeoutMs = Number(argValue("--timeout-ms", String(requestTimeoutMs)));
  requestRetries = Number(argValue("--retries", String(requestRetries)));
  requestDelayMs = Number(argValue("--delay-ms", "0"));
  const pageOverride = argValue("--pages", null);
  fs.mkdirSync(CACHE, { recursive: true });

  const existing = loadExistingDetails();
  if (existing.size) {
    const rows = [...existing.entries()].sort((a, b) => a[0] - b[0]).map(([, value]) => value);
    writeDurableJson(OUT, SNAPSHOT_OUT, rows);
  }
  if (hasArg("--prune-only")) {
    console.log(`pruned cache: items=${existing.size}`);
    return;
  }

  const items = hasArg("--reuse-list") && (fs.existsSync(LIST_CACHE) || fs.existsSync(SNAPSHOT_LIST))
    ? loadJson(LIST_CACHE, loadJson(SNAPSHOT_LIST, []))
    : await collectList(pageOverride ? Number(pageOverride) : null);
  if (items.length) writeDurableJson(LIST_CACHE, SNAPSHOT_LIST, items);

  const details = loadExistingDetails();
  const todo = items.filter(item => {
    const row = details.get(Number(item.id));
    return !row || row.error || !row.tooltip;
  });
  console.log(`detail queue: ${todo.length}/${items.length}`);

  let done = 0;
  let failed = 0;
  await runQueue(todo, workers, async (item, _index, workerIndex) => {
    let row;
    try {
      await politeDelay(workerIndex);
      row = await fetchDetail(item);
      details.set(Number(row.id), row);
    } catch (error) {
      failed += 1;
    }
    done += 1;
    if (done % 100 === 0 || done === todo.length) {
      const rows = [...details.entries()].sort((a, b) => a[0] - b[0]).map(([, value]) => value);
      writeDurableJson(OUT, SNAPSHOT_OUT, rows);
      console.log(`detail ${done}/${todo.length} complete=${details.size}/${items.length} failed=${failed}`);
    }
  });

  const rows = [...details.entries()].sort((a, b) => a[0] - b[0]).map(([, value]) => value);
  writeDurableJson(OUT, SNAPSHOT_OUT, rows);
  const greenItems = rows.filter(row => row.green && row.green.length).length;
  console.log(`wrote ${OUT}`);
  console.log(`items=${rows.length} green_items=${greenItems} pending=${items.length - rows.length}`);
}

main().catch(error => {
  console.error(error);
  process.exit(1);
});
